tests/TryCatchTest.php

```diff
@@ -491,6 +491,613 @@ function foo() : void {
+                        $bar = 3;
+                    }
+                ',
+                'assertions' => [
+                    '$foo===' => '3',
+                    '$bar===' => '3',
+                ],
+            ],
+            'universalCatchNarrowsTypeInFinally' => [
+                'code' => '<?php
+                    try {
+                        $var = 1;
+                        $var = 2;
+                        $var = 3;
+                    } catch (Exception $_) {
+                        $var = 4;
+                        $var = 5;
+                    } catch (Throwable $_) {
+                        $var = 6;
+                        $var = 7;
+                    } finally {
+                        // Since an exception will always be caught and $var will be changed, $var can only be 3|5|7.
+                        /** @psalm-check-type $var = 3|5|7 */;
+                    }
+                ',
+                'assertions' => [
+                    '$var===' => '3|5|7',
+                ]
+            ],
+            'nonUniversalCatchDoesntNarrowTypeInFinally' => [
+                'code' => '<?php
+                    try {
+                        $var = 1;
+                        $var = 2;
+                        $var = 3;
+                    } catch (InvalidArgumentException $_) {
+                        $var = 4;
+                        $var = 5;
+                    } catch (RuntimeException $_) {
+                        $var = 6;
+                        $var = 7;
+                    } finally {
+                        // $var could be 1 or 2 if SomeOtherException is thrown
+                        /** @psalm-check-type $var? = 1|2|3|5|7 */;
+                    }
+                ',
+                'assertions' => [
+                    '$var===' => '3|5|7',
+                ]
+            ],
+            'createByRefInIfInTry' => [
+                'code' => '<?php
+                    /**
+                     * @param mixed $var
+                     * @param-out int $var
+                     */
+                    function assignToInt(&$var): void
+                    {
+                        $var = 1;
+                    }
+
+                    try {
+                        if (assignToInt($test)) {
+                        }
+                    } catch (Exception $_) {
+                    }
+                ',
+                'assertions' => [
+                    '$test?' => 'int',
+                ],
+            ],
+            'extractInTry' => [
+                'code' => '<?php
+                    $foo = 1;
+                    $bar = "baz";
+                    try {
+                        extract([]); // extract overrides all variables with `mixed`.
+                    } catch (Exception $_) {
+                    }
+                ',
+                'assertions' => [
+                    '$foo' => 'int|mixed',
+                    '$bar' => 'mixed|string',
```

Comment:
We could probably add a TypeCombiner::combine to collapse those

Reply:
> the idea was to allow better completion

I think it's just a bug. Given the sheer amount of Unions created, it would probably be a horrible place to put an automatic combining there for performances, but IMHO, we should never end up with this kind of types, it's just the symptom of a wrong handling somewhere